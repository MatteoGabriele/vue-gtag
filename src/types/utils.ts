export type Tail<T> = T extends [infer _, ...infer U] ? U : never;
