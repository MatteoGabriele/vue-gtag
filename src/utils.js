export const load = (url, options = {}) => {
  return new Promise((resolve, reject) => {
    if (typeof document === "undefined") {
      return;
    }

    const head = document.head || document.getElementsByTagName("head")[0];
    const script = document.createElement("script");

    script.async = true;
    script.src = url;
    script.defer = options.defer;

    if (options.preconnectOrigin) {
      const link = document.createElement("link");

      link.href = options.preconnectOrigin;
      link.rel = "preconnect";

      head.appendChild(link);
    }

    head.appendChild(script);

    script.onload = resolve;
    script.onerror = reject;
  });
};

export const isFn = (fn) => typeof fn === "function";

export const isObject = (item) => {
  return item && typeof item === "object" && !Array.isArray(item);
};

export const mergeDeep = (target, ...sources) => {
  if (!sources.length) {
    return target;
  }

  const source = sources.shift();

  if (!isObject(target) || !isObject(source)) {
    return;
  }

  for (const key in source) {
    if (isObject(source[key])) {
      if (!target[key]) {
        Object.assign(target, { [key]: {} });
      }

      mergeDeep(target[key], source[key]);
    } else {
      Object.assign(target, { [key]: source[key] });
    }
  }

  return mergeDeep(target, ...sources);
};

export const isBrowser = () => {
  if (typeof window === "undefined" || typeof document === "undefined") {
    return false;
  }

  return true;
};

export const warn = (text, shouldLog = true) => {
  if (!isBrowser() || process.env.NODE_ENV === "production") {
    return;
  }

  if (!shouldLog) {
    return;
  }

  console.warn(`[vue-gtag] ${text}`);
};
