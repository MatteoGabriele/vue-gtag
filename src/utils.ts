export async function injectScript(
  url: string,
  options?: {
    preconnectOrigin?: string;
    defer?: boolean;
  },
): Promise<void> {
  return new Promise((resolve, reject) => {
    if (typeof document === "undefined") {
      return;
    }

    const head = document.head || document.getElementsByTagName("head")[0];
    const script = document.createElement("script");

    script.async = true;
    script.src = url;

    if (options?.defer) {
      script.defer = true;
    }

    if (options?.preconnectOrigin) {
      const link = document.createElement("link");

      link.href = options.preconnectOrigin;
      link.rel = "preconnect";

      head.appendChild(link);
    }

    head.appendChild(script);

    script.onload = () => resolve;
    script.onerror = reject;
  });
}
