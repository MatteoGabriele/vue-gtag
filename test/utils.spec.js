import * as util from "@/utils";
import flushPromises from "flush-promises";

describe("load", () => {
  afterEach(() => {
    document.getElementsByTagName("html")[0].innerHTML = "";
  });

  it("should return a promise", () => {
    expect(util.load("a")).toBeInstanceOf(Promise);
  });

  it("should create a script tag", async () => {
    util.load("foo");

    await flushPromises();

    expect(document.head).toMatchSnapshot();
  });

  it("should create a link for domain preconnect", async () => {
    util.load("foo", {
      preconnectOrigin: "bar",
    });

    await flushPromises();

    expect(document.head).toMatchSnapshot();
  });

  it("should create a script tag with the defer attribute", async () => {
    util.load("foo", {
      defer: true,
    });

    await flushPromises();

    expect(document.head).toMatchSnapshot();
  });
});

describe("mergeDeep", () => {
  it("should merge two objects", () => {
    const a = { a: 1, c: { a: 1 } };
    const b = { b: 1, c: { b: 1 } };
    const c = util.mergeDeep(a, b);

    expect(c).toMatchObject({
      a: 1,
      b: 1,
      c: {
        b: 1,
        a: 1,
      },
    });
  });

  it("should merge objects containing arrays", () => {
    const a = { c: [1, 2, 3] };
    const b = { c: [4, 5], d: { b: 1 } };
    const c = util.mergeDeep(a, b);

    expect(c).toEqual({
      c: [4, 5],
      d: { b: 1 },
    });
  });
});

describe("warn", () => {
  it("should warn with a customized message prefix", () => {
    console.warn = jest.fn();
    util.warn("foo");
    expect(console.warn).toHaveBeenCalledWith("[vue-gtag] foo");
  });
});
