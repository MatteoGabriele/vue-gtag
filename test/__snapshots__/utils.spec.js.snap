// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load should create a link for domain preconnect 1`] = `
<head>
  <link
    href="bar"
    rel="preconnect"
  />
  <script
    src="foo"
  />
</head>
`;

exports[`load should create a script tag 1`] = `
<head>
  <script
    src="foo"
  />
</head>
`;

exports[`load should create a script tag with the defer attribute 1`] = `
<head>
  <script
    defer=""
    src="foo"
  />
</head>
`;
